src/wasm32.rs

```diff
@@ -0,0 +1,563 @@
+// traits for cells instead of bounding the inner / array types in a separate trait. This makes
+// Cells redundant for stores (due to Cell::get_mut) but makes the list of exposed loads more
+// terse.
+//
+// As we mark the cell type, and arrays of such cells, the specific underlying types must be chosen
+// to avoid overlapping impls. This is a dance around the type checker as the many recursive impls
+// have potential of infinitely recursing until the next generation. Hence we make sure to always
+// reduce the matched type: both [Cell<T>; N] and Cell<T> take out a cell without introducing more
+// types to the obligations. E.g. Cell<[T; N]> for [T: N]  would not work.
+
+impl<T, const N: usize> Is1ByteUnaligned for [core::cell::Cell<T>; N] where [T; N]: Is1ByteUnaligned {}
+
+impl<T> Is1ByteUnaligned for core::cell::Cell<T> where T: Is1ByteUnaligned {}
+
+impl<T, const N: usize> Is2BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is2BytesUnaligned
+{
+}
+
+impl<T> Is2BytesUnaligned for core::cell::Cell<T> where T: Is2BytesUnaligned {}
+
+impl<T, const N: usize> Is4BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is4BytesUnaligned
+{
+}
+
+impl<T> Is4BytesUnaligned for core::cell::Cell<T> where T: Is4BytesUnaligned {}
+
+impl<T, const N: usize> Is8BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is8BytesUnaligned
+{
+}
+
+impl<T> Is8BytesUnaligned for core::cell::Cell<T> where T: Is8BytesUnaligned {}
+
+impl<T, const N: usize> Is16BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is16BytesUnaligned
+{
+}
+
+impl<T> Is16BytesUnaligned for core::cell::Cell<T> where T: Is16BytesUnaligned {}
+
+#[cfg(test)]
+mod tests {
+    use core::arch::wasm32::{self as arch, v128};
+
+    fn assert_v128_bytes<const N: usize>(val: v128, data: &[[u8; N]]) {
+        assert_eq!(
+            unsafe { core::mem::transmute::<_, [u8; 16]>(val) },
+            unsafe { core::slice::from_raw_parts(data.as_ptr() as *const u8, data.len() * N) }
+        );
+
+        // assert_eq!(arch::to_bytes(val), <[u8; N]>::flat(data))
+    }
+
+    #[test]
+    #[cfg_attr(not(target_feature = "simd128"), ignore)]
+    fn test_i16x8_load_extend_i8x8() {
+        #[target_feature(enable = "simd128")]
+        fn test(a: [i8; 8]) {
+            let v = super::i16x8_load_extend_i8x8(&a);
+            assert_v128_bytes(v, &a.map(|i| (i as i16).to_ne_bytes()));
+        }
+
+        let a = core::array::from_fn(|i| i as i8);
+        test(a)
+    }
+
+    #[test]
+    #[cfg_attr(not(target_feature = "simd128"), ignore)]
+    pub fn test_i16x8_load_extend_u8x8() {
+        #[target_feature(enable = "simd128")]
+        fn test(a: [u8; 8]) {
+            let v = super::u16x8_load_extend_u8x8(&a);
+            assert_v128_bytes(v, &a.map(|i| (i as i16).to_ne_bytes()));
+        }
+
+        let a = core::array::from_fn(|i| u8::MAX - i as u8);
```

Comment:
Could you add a comment stating the `uNxN` tests are checking for zero-extension?

We do `u8:MAX - i`, to fill the top bit and make sure it's not sign-extending.